File: 06/validate credit card/validate.py
from cs50 import get_int

# Validate credit card using Luhn’s Algorithm and determine whether it's a VISA, Mastercard or Amex

# SOLUTION

# Get card number from user
number = get_int("Number: ")

# Find the length of the card number
n = len(str(number))

divide = 1
total_second = 0

# Loop through every second digit and apply the algorithm
for i in range(round(n / 2)):
    divide = divide * 100
    second_number = number % divide
    second_number = int(second_number / (divide / 10))
    multiplied = second_number * 2
    if (multiplied >= 10):
        first = int(multiplied / 10)
        second = multiplied % 10
        total_second = total_second + first + second
    else:
        total_second = total_second + multiplied

not_divide = 0.1
total_first = 0
x = 1

# Loop through every first digit and apply the algorithm
for j in range(round(n / 2)):
    not_divide = not_divide * 100
    first_number = number % not_divide
    first_number = int(first_number / x)
    x = x * 100
    total_first = total_first + first_number

# Calculate the total sum
total_sum = total_second + total_first


# If the card is invalid print INVALID
if (total_sum % 10 != 0):
    print("INVALID")

# Find which card is the valid card
else:
    if (int(n) == 15 and (int(number / 10000000000000) == 34 or int(number / 10000000000000) == 37)):
        print("AMEX")
    elif (int(n) == 16 and (int(number / 100000000000000) >= 51 and int(number / 100000000000000) <= 55)):
        print("MASTERCARD")
    elif ((int(n) == 16 or int(n) == 13) and (int(number / 1000000000000000) == 4 or int(number / 1000000000000) == 4)):
        print("VISA")

File: 01/validate credit card/validate.c
#include <cs50.h>
#include <stdio.h>

// Validate credit card using Luhn’s Algorithm and determine whether it's a VISA, Mastercard or Amex

//SOLUTION

int main(void)
{

    //Get user input
    long int creditCardNumber = get_long("Number:");
    long int creditCardNumberCount = creditCardNumber;

    //Find the number of digits in the creditCardNumber
    int count = 0;
    while (creditCardNumberCount != 0)
    {
        creditCardNumberCount /= 10;
        count++;
    }

    long int remainderOne;
    long int sumSecondDigits = 0;
    long int x = 1;
    double y = 0.1;
    int secondDigit;
    int productDigit;
    int separatedRemainder;
    int separatedDigit;
    int sumProductDigit;

    // Calculate the sum of every second digit in the creditCardNumber
    for (int i = 0; i < count / 2; i++)
    {
        x = x * 100;
        y = y * 100;

        remainderOne = creditCardNumber % x;
        secondDigit = remainderOne / y;
        productDigit = secondDigit * 2;

        // Separate all double digits into single and add them together
        if (productDigit >= 10)
        {
            separatedRemainder = productDigit % 10;
            separatedDigit = productDigit / 10;
            sumProductDigit = separatedRemainder + separatedDigit;

            //Calculate the sum of all second digits if there are double digits
            sumSecondDigits = sumSecondDigits + sumProductDigit;
        }
        else
        {
            //Calculate the sum of all second digits if there are no double digits
            sumSecondDigits = sumSecondDigits + productDigit;
        }
    }

    double z = 0.1;
    double a = 0.01;
    long int remainderTwo;
    int firstDigit;
    int sumFirstDigits = 0;

    // Calculate the sum of the remaining digits in the creditCardNumber
    for (int i = 0; i <= count / 2; i++)
    {
        z = z * 100;
        a = a * 100;
        long int b = z;

        remainderTwo = creditCardNumber % b;
        firstDigit = remainderTwo / a;
        sumFirstDigits = sumFirstDigits + firstDigit;
    }

    //Calculate the total sum
    int sum;
    sum = sumSecondDigits + sumFirstDigits;

    //Get the first two or 1 digit for each type of credit card
    int AmericanExpress = creditCardNumber / 10000000000000;
    int MasterCard = creditCardNumber / 100000000000000;
    int Visa13 = creditCardNumber / 1000000000000;
    int Visa16 = creditCardNumber / 1000000000000000;

    //Print INVALID if credit card number provided is not valid
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (count == 15 && (AmericanExpress == 34 || AmericanExpress == 37))
    {
        printf("AMEX\n");
    }

    else if (count == 16 && (MasterCard > 50 && MasterCard < 56))
    {
        printf("MASTERCARD\n");
    }

    else if ((count == 13 || count == 16) && (Visa13 == 4 || Visa16 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}